Flacky Compose screen test

**Describe the bug**
AuthComposeActivity Test sometimes fails.

**Expected behavior**
The test should always behave the same way

**Additional context**
Failed on Emulator 30
Logs from CI:
```kotlin
org.fnives.test.showcase.ui.AuthComposeInstrumentedTest > networkErrorShowsProperErrorMessage[test(AVD) - 11] FAILED 
	androidx.compose.ui.test.ComposeTimeoutException: Condition still not satisfied after 1000 ms
	at androidx.compose.ui.test.junit4.AbstractMainTestClock$advanceTimeUntil$1.invoke(AbstractMainTestClock.kt:58)
org.fnives.test.showcase.ui.login.codekata.CodeKataAuthActivitySharedTest > null[test(AVD) - 11] SKIPPED 
Tests on test(AVD) - 11 failed: There was 1 failure(s).
Jul 12, 2022 7:26:56 AM com.google.testing.platform.RunnerImpl$Companion summarizeAndLogTestResult
SEVERE: Execute org.fnives.test.showcase.ui.AuthComposeInstrumentedTest.networkErrorShowsProperErrorMessage: FAILED
androidx.compose.ui.test.ComposeTimeoutException: androidx.compose.ui.test.ComposeTimeoutException: Condition still not satisfied after 1000 ms
at androidx.compose.ui.test.junit4.AbstractMainTestClock$advanceTimeUntil$1.invoke(AbstractMainTestClock.kt:58)
at androidx.compose.ui.test.junit4.AbstractMainTestClock$advanceTimeUntil$1.invoke(AbstractMainTestClock.kt:54)
at androidx.compose.ui.test.junit4.AndroidSynchronization_androidKt.runOnUiThread$lambda-0(AndroidSynchronization.android.kt:37)
at androidx.compose.ui.test.junit4.AndroidSynchronization_androidKt.$r8$lambda$JHfZTn-RG8lrYeP9BZG_AgROyug(Unknown Source:0)
at androidx.compose.ui.test.junit4.AndroidSynchronization_androidKt$$ExternalSyntheticLambda0.call(Unknown Source:2)
at java.util.concurrent.FutureTask.run(FutureTask.java:[266](https://github.com/fknives/AndroidTest-ShowCase/runs/7296766659?check_suite_focus=true#step:7:270))
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at android.app.Instrumentation$SyncRunnable.run(Instrumentation.java:2223)
at android.os.Handler.handleCallback(Handler.java:938)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loop(Looper.java:223)
at android.app.ActivityThread.main(ActivityThread.java:7656)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
androidx.compose.ui.test.ComposeTimeoutException: Condition still not satisfied after 1000 ms
at androidx.compose.ui.test.junit4.AbstractMainTestClock$advanceTimeUntil$1.invoke(AbstractMainTestClock.kt:58)
at androidx.compose.ui.test.junit4.AbstractMainTestClock$advanceTimeUntil$1.invoke(AbstractMainTestClock.kt:54)
at androidx.compose.ui.test.junit4.AndroidSynchronization_androidKt.runOnUiThread$lambda-0(AndroidSynchronization.android.kt:37)
at androidx.compose.ui.test.junit4.AndroidSynchronization_androidKt.$r8$lambda$JHfZTn-RG8lrYeP9BZG_AgROyug(Unknown Source:0)
at androidx.compose.ui.test.junit4.AndroidSynchronization_androidKt$$ExternalSyntheticLambda0.call(Unknown Source:2)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at android.app.Instrumentation$SyncRunnable.run(Instrumentation.java:2223)
at android.os.Handler.handleCallback(Handler.java:938)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loop(Looper.java:223)
at android.app.ActivityThread.main(ActivityThread.java:7656)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
Test results saved as file:/Users/runner/work/AndroidTest-ShowCase/AndroidTest-ShowCase/app/build/outputs/androidTest-results/connected/test(AVD)%20-%2011/test-result.pb. Inspect these results in Android Studio by selecting Run > Import Tests From File from the menu bar and importing test-result.pb.
> Task :app:connectedDebugAndroidTest FAILED
FAILURE: Build failed with an exception.
```

Similar thing also happened on API 21
```kotlin
org.fnives.test.showcase.ui.AuthComposeInstrumentedTest > emptyUserNameShowsProperErrorMessage[test(AVD) - 5.0.2] FAILED 
	java.lang.AssertionError: Failed to assert the following: (Text + EditableText contains 'Username is not filled properly!' (ignoreCase: false))
	Reason: Expected exactly '1' node but could not find any node that satisfies: (TestTag = 'AuthScreenTag.LoginError')
org.fnives.test.showcase.ui.login.codekata.CodeKataAuthActivitySharedTest > null[test(AVD) - 5.0.2] SKIPPED 
Tests on test(AVD) - 5.0.2 failed: There was 1 failure(s).
Jul 12, 2022 11:37:18 AM com.google.testing.platform.RunnerImpl$Companion summarizeAndLogTestResult
SEVERE: Execute org.fnives.test.showcase.ui.AuthComposeInstrumentedTest.emptyUserNameShowsProperErrorMessage: FAILED
java.lang.AssertionError: java.lang.AssertionError: Failed to assert the following: (Text + EditableText contains 'Username is not filled properly!' (ignoreCase: false))
Reason: Expected exactly '1' node but could not find any node that satisfies: (TestTag = 'AuthScreenTag.LoginError')
at androidx.compose.ui.test.SemanticsNodeInteraction.fetchOneOrDie(SemanticsNodeInteraction.kt:162)
at androidx.compose.ui.test.SemanticsNodeInteraction.fetchSemanticsNode(SemanticsNodeInteraction.kt:96)
at androidx.compose.ui.test.AssertionsKt.assert(Assertions.kt:268)
at androidx.compose.ui.test.AssertionsKt.assert$default(Assertions.kt:260)
at androidx.compose.ui.test.AssertionsKt.assertTextContains(Assertions.kt:214)
at androidx.compose.ui.test.AssertionsKt.assertTextContains$default(Assertions.kt:209)
at org.fnives.test.showcase.ui.ComposeLoginRobot.assertErrorIsShown(ComposeLoginRobot.kt:50)
at org.fnives.test.showcase.ui.AuthComposeInstrumentedTest.emptyUserNameShowsProperErrorMessage(AuthComposeInstrumentedTest.kt:100)
java.lang.AssertionError: Failed to assert the following: (Text + EditableText contains 'Username is not filled properly!' (ignoreCase: false))
Reason: Expected exactly '1' node but could not find any node that satisfies: (TestTag = 'AuthScreenTag.LoginError')
at androidx.compose.ui.test.SemanticsNodeInteraction.fetchOneOrDie(SemanticsNodeInteraction.kt:162)
at androidx.compose.ui.test.SemanticsNodeInteraction.fetchSemanticsNode(SemanticsNodeInteraction.kt:96)
at androidx.compose.ui.test.AssertionsKt.assert(Assertions.kt:268)
at androidx.compose.ui.test.AssertionsKt.assert$default(Assertions.kt:260)
at androidx.compose.ui.test.AssertionsKt.assertTextContains(Assertions.kt:214)
at androidx.compose.ui.test.AssertionsKt.assertTextContains$default(Assertions.kt:209)
at org.fnives.test.showcase.ui.ComposeLoginRobot.assertErrorIsShown(ComposeLoginRobot.kt:50)
at org.fnives.test.showcase.ui.AuthComposeInstrumentedTest.emptyUserNameShowsProperErrorMessage(AuthComposeInstrumentedTest.kt:100)
```

--- a/app/src/androidTest/java/org/fnives/test/showcase/ui/AuthComposeInstrumentedTest.kt
+++ b/app/src/androidTest/java/org/fnives/test/showcase/ui/AuthComposeInstrumentedTest.kt
@@ -1,11 +1,16 @@
 package org.fnives.test.showcase.ui
 
+import androidx.compose.ui.test.MainTestClock
 import androidx.compose.ui.test.junit4.StateRestorationTester
 import androidx.compose.ui.test.junit4.createComposeRule
+import androidx.test.espresso.IdlingRegistry
+import androidx.test.espresso.IdlingResource
 import androidx.test.ext.junit.runners.AndroidJUnit4
 import org.fnives.test.showcase.R
 import org.fnives.test.showcase.android.testutil.screenshot.ScreenshotRule
-import org.fnives.test.showcase.android.testutil.synchronization.idlingresources.anyResourceIdling
+import org.fnives.test.showcase.android.testutil.synchronization.idlingresources.anyResourceNotIdle
+import org.fnives.test.showcase.android.testutil.synchronization.idlingresources.awaitUntilIdle
+import org.fnives.test.showcase.android.testutil.synchronization.loopMainThreadFor
 import org.fnives.test.showcase.compose.screen.AppNavigation
 import org.fnives.test.showcase.core.integration.fake.FakeUserDataLocalStorage
 import org.fnives.test.showcase.core.login.IsUserLoggedInUseCase
@@ -18,6 +23,7 @@ import org.junit.Test
 import org.junit.rules.RuleChain
 import org.junit.runner.RunWith
 import org.koin.test.KoinTest
+import java.util.concurrent.Executors
 
 @RunWith(AndroidJUnit4::class)
 class AuthComposeInstrumentedTest : KoinTest {
@@ -53,8 +59,8 @@ class AuthComposeInstrumentedTest : KoinTest {
         mockServerScenarioSetup.setScenario(
             AuthScenario.Success(password = "alma", username = "banan")
         )
+        composeTestRule.mainClock.advanceTimeBy(SPLASH_DELAY)
 
-        composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
         navigationRobot.assertAuthScreen()
         robot.setPassword("alma")
             .setUsername("banan")
@@ -67,21 +73,21 @@ class AuthComposeInstrumentedTest : KoinTest {
         robot.assertLoading()
         composeTestRule.mainClock.autoAdvance = true
 
-        composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+        composeTestRule.mainClock.awaitIdlingResources()
         navigationRobot.assertHomeScreen()
     }
 
     /** GIVEN empty password and username WHEN signIn THEN error password is shown */
     @Test
     fun emptyPasswordShowsProperErrorMessage() {
-        composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+        composeTestRule.mainClock.advanceTimeBy(SPLASH_DELAY)
         navigationRobot.assertAuthScreen()
 
         robot.setUsername("banan")
             .assertUsername("banan")
             .clickOnLogin()
 
-        composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+        composeTestRule.mainClock.awaitIdlingResources()
         robot.assertErrorIsShown(R.string.password_is_invalid)
             .assertNotLoading()
         navigationRobot.assertAuthScreen()
@@ -90,7 +96,7 @@ class AuthComposeInstrumentedTest : KoinTest {
     /** GIVEN password and empty username WHEN signIn THEN error username is shown */
     @Test
     fun emptyUserNameShowsProperErrorMessage() {
-        composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+        composeTestRule.mainClock.advanceTimeBy(SPLASH_DELAY)
         navigationRobot.assertAuthScreen()
 
         robot
@@ -98,7 +104,7 @@ class AuthComposeInstrumentedTest : KoinTest {
             .assertPassword("banan")
             .clickOnLogin()
 
-        composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+        composeTestRule.mainClock.awaitIdlingResources()
         robot.assertErrorIsShown(R.string.username_is_invalid)
             .assertNotLoading()
         navigationRobot.assertAuthScreen()
@@ -111,7 +117,7 @@ class AuthComposeInstrumentedTest : KoinTest {
             AuthScenario.InvalidCredentials(password = "alma", username = "banan")
         )
 
-        composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+        composeTestRule.mainClock.advanceTimeBy(SPLASH_DELAY)
         navigationRobot.assertAuthScreen()
         robot.setUsername("alma")
             .setPassword("banan")
@@ -124,7 +130,7 @@ class AuthComposeInstrumentedTest : KoinTest {
         robot.assertLoading()
         composeTestRule.mainClock.autoAdvance = true
 
-        composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+        composeTestRule.mainClock.awaitIdlingResources()
         robot.assertErrorIsShown(R.string.credentials_invalid)
             .assertNotLoading()
         navigationRobot.assertAuthScreen()
@@ -137,7 +143,7 @@ class AuthComposeInstrumentedTest : KoinTest {
             AuthScenario.GenericError(username = "alma", password = "banan")
         )
 
-        composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+        composeTestRule.mainClock.advanceTimeBy(SPLASH_DELAY)
         navigationRobot.assertAuthScreen()
         robot.setUsername("alma")
             .setPassword("banan")
@@ -150,7 +156,7 @@ class AuthComposeInstrumentedTest : KoinTest {
         robot.assertLoading()
         composeTestRule.mainClock.autoAdvance = true
 
-        composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+        composeTestRule.mainClock.awaitIdlingResources()
         robot.assertErrorIsShown(R.string.something_went_wrong)
             .assertNotLoading()
         navigationRobot.assertAuthScreen()
@@ -159,7 +165,7 @@ class AuthComposeInstrumentedTest : KoinTest {
     /** GIVEN username and password WHEN restoring THEN username and password fields contain the same text */
     @Test
     fun restoringContentShowPreviousCredentials() {
-        composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+        composeTestRule.mainClock.advanceTimeBy(SPLASH_DELAY)
         navigationRobot.assertAuthScreen()
         robot.setUsername("alma")
             .setPassword("banan")
@@ -171,5 +177,37 @@ class AuthComposeInstrumentedTest : KoinTest {
         navigationRobot.assertAuthScreen()
         robot.assertUsername("alma")
             .assertPassword("banan")
+    }
+
+    companion object {
+        private const val SPLASH_DELAY = 600L
+
+        // workaround, issue with idlingResources is tracked here https://github.com/robolectric/robolectric/issues/4807
+        /**
+         * Await the idling resource on a different thread while looping main.
+         */
+        fun MainTestClock.awaitIdlingResources() {
+            val idlingRegistry = IdlingRegistry.getInstance()
+            if (!anyResourceNotIdle()) return
+
+            val executor = Executors.newSingleThreadExecutor()
+            var isIdle = false
+            executor.submit {
+                do {
+                    idlingRegistry.resources
+                        .filterNot(IdlingResource::isIdleNow)
+                        .forEach { idlingResource ->
+                            idlingResource.awaitUntilIdle()
+                        }
+                } while (!idlingRegistry.resources.all(IdlingResource::isIdleNow))
+                isIdle = true
+            }
+            while (!isIdle) {
+                loopMainThreadFor(200L)
+            }
+            executor.shutdown()
+
+            advanceTimeByFrame()
+        }
     }
 }

--- a/app/src/sharedTest/java/org/fnives/test/showcase/testutils/idling/DatabaseDispatcherTestRule.kt
+++ b/app/src/sharedTest/java/org/fnives/test/showcase/testutils/idling/DatabaseDispatcherTestRule.kt
@@ -3,7 +3,7 @@ package org.fnives.test.showcase.testutils.idling
 import kotlinx.coroutines.ExperimentalCoroutinesApi
 import kotlinx.coroutines.test.StandardTestDispatcher
 import kotlinx.coroutines.test.TestDispatcher
-import org.fnives.test.showcase.android.testutil.synchronization.idlingresources.anyResourceIdling
+import org.fnives.test.showcase.android.testutil.synchronization.idlingresources.anyResourceNotIdle
 import org.fnives.test.showcase.android.testutil.synchronization.idlingresources.awaitIdlingResources
 import org.fnives.test.showcase.android.testutil.synchronization.runOnUIAwaitOnCurrent
 import org.fnives.test.showcase.testutils.storage.TestDatabaseInitialization
@@ -42,7 +42,7 @@ class DatabaseDispatcherTestRule : TestRule {
     companion object {
         fun TestDispatcher.advanceUntilIdleWithIdlingResources() {
             scheduler.advanceUntilIdle() // advance until a request is sent
-            while (anyResourceIdling()) { // check if any request is in progress
+            while (anyResourceNotIdle()) { // check if any request is in progress
                 awaitIdlingResources() // complete all requests and other idling resources
                 scheduler.advanceUntilIdle() // run coroutines after request is finished
             }

--- a/codekata/compose.instructionset.md
+++ b/codekata/compose.instructionset.md
@@ -159,7 +159,7 @@ private val mockServerScenarioSetup get() = mockServerScenarioSetupTestRule.mock
 Coroutine setup is the same, except for `Dispatchers.setMain(dispatcher)`, which we don't need. 
 
 ```kotlin
-private val dispatcherTestRule = DispatcherTestRule()
+private val dispatcherTestRule = DatabaseDispatcherTestRule()
 ```
 
 Setting the rules:
@@ -182,12 +182,10 @@ mockServerScenarioSetup.setScenario(
 )
 ```
 
-Then we wait for the idling resources, more precisely for the app to navigate us correctly to AuthScreen since we're not logged in:
+Then we wait a bit, more precisely we wait for the app to navigate us correctly to AuthScreen since we're not logged in:
 ```kotlin
-composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+composeTestRule.mainClock.advanceTimeBy(510L)
 ```
-
-> Note: Considering what the docs say this shouldn't be necessarily if the idling resources are setup in Espresso, since the compose test rule is aware of espresso and it waits for idle before every finder. In practice it only works with the line above. Could be a bug somewhere.  
 
 We assert that we are indeed on the correct screen
 ```kotlin
@@ -216,9 +214,12 @@ composeTestRule.mainClock.autoAdvance = true // Let clock auto advance again
 
 Lastly we check the navigation was correct, meaning we should be on the home screen:
 ```kotlin
-composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() } // wait for login network call
+composeTestRule.mainClock.awaitIdlingResources() // wait for login network call idling resource
 navigationRobot.assertHomeScreen()
 ```
+
+> `awaitIdlingResources` is an extension function to await all idling resources.
+> Note: Considering what the docs say this shouldn't be necessarily if the idling resources are setup in Espresso, since the compose test rule is aware of espresso and it waits for idle before every finder. In practice it only works with the line above. Could be a bug somewhere.
 
 ### 2. `emptyPasswordShowsProperErrorMessage`
 
@@ -226,7 +227,7 @@ Next up we verify what happens if the user doesn't set their password. We don't 
 
 First we check that we are in the write place:
 ```kotlin
-composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+composeTestRule.mainClock.advanceTimeBy(510L)
 navigationRobot.assertAuthScreen()
 ```
 
@@ -239,7 +240,7 @@ robot.setUsername("banan")
 
 Finally we let coroutines go and verify the error is shown and we have not navigated:
 ```kotlin
-composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+composeTestRule.mainClock.awaitIdlingResources()
 robot.assertErrorIsShown(R.string.password_is_invalid)
     .assertNotLoading()
 navigationRobot.assertAuthScreen()
@@ -251,7 +252,7 @@ This will be really similar as the previous test, so try to do it on your own. T
 
 Still, here is the complete code:
 ```kotlin
-composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+composeTestRule.mainClock.advanceTimeBy(510L)
 navigationRobot.assertAuthScreen()
 
 robot
@@ -259,7 +260,7 @@ robot
     .assertPassword("banan")
     .clickOnLogin()
 
-composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+composeTestRule.mainClock.awaitIdlingResources()
 robot.assertErrorIsShown(R.string.username_is_invalid)
     .assertNotLoading()
 navigationRobot.assertAuthScreen()
@@ -276,7 +277,7 @@ mockServerScenarioSetup.setScenario(
 
 Now input the credentials and fire the event:
 ```kotlin
-composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+composeTestRule.mainClock.advanceTimeBy(510L)
 navigationRobot.assertAuthScreen()
 robot.setUsername("alma")
     .setPassword("banan")
@@ -292,7 +293,7 @@ composeTestRule.mainClock.autoAdvance = true
 
 Now at the end verify the error is shown properly:
 ```kotlin
-composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+composeTestRule.mainClock.awaitIdlingResources()
 robot.assertErrorIsShown(R.string.credentials_invalid)
     .assertNotLoading()
 navigationRobot.assertAuthScreen()
@@ -309,7 +310,7 @@ mockServerScenarioSetup.setScenario(
     AuthScenario.GenericError(username = "alma", password = "banan")
 )
 
-composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+composeTestRule.mainClock.advanceTimeBy(510L)
 navigationRobot.assertAuthScreen()
 robot.setUsername("alma")
     .setPassword("banan")
@@ -322,7 +323,7 @@ composeTestRule.mainClock.advanceTimeByFrame()
 robot.assertLoading()
 composeTestRule.mainClock.autoAdvance = true
 
-composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+composeTestRule.mainClock.awaitIdlingResources()
 robot.assertErrorIsShown(R.string.something_went_wrong)
     .assertNotLoading()
 navigationRobot.assertAuthScreen()
@@ -342,7 +343,7 @@ Then in `setup()`, we need to `setContent` on `stateRestorationTester` instead o
 Now for the actual test, we first setup the content then we trigger restoration by calling `stateRestorationTester.emulateSavedInstanceStateRestore()`, afterwards we can verify that the content is recreated in the correct way:
 
 ```kotlin
-composeTestRule.mainClock.advanceTimeUntil { anyResourceIdling() }
+composeTestRule.mainClock.advanceTimeBy(510L)
 navigationRobot.assertAuthScreen()
 robot.setUsername("alma")
     .setPassword("banan")

--- a/test-util-android/src/main/java/org/fnives/test/showcase/android/testutil/synchronization/MainDispatcherTestRule.kt
+++ b/test-util-android/src/main/java/org/fnives/test/showcase/android/testutil/synchronization/MainDispatcherTestRule.kt
@@ -7,7 +7,7 @@ import kotlinx.coroutines.test.TestDispatcher
 import kotlinx.coroutines.test.UnconfinedTestDispatcher
 import kotlinx.coroutines.test.resetMain
 import kotlinx.coroutines.test.setMain
-import org.fnives.test.showcase.android.testutil.synchronization.idlingresources.anyResourceIdling
+import org.fnives.test.showcase.android.testutil.synchronization.idlingresources.anyResourceNotIdle
 import org.fnives.test.showcase.android.testutil.synchronization.idlingresources.awaitIdlingResources
 import org.junit.rules.TestRule
 import org.junit.runner.Description
@@ -54,7 +54,7 @@ open class MainDispatcherTestRule(private val useStandard: Boolean = true) : Tes
     companion object {
         fun TestDispatcher.advanceUntilIdleWithIdlingResources() {
             scheduler.advanceUntilIdle() // advance until a request is sent
-            while (anyResourceIdling()) { // check if any request is in progress
+            while (anyResourceNotIdle()) { // check if any request is in progress
                 awaitIdlingResources() // complete all requests and other idling resources
                 scheduler.advanceUntilIdle() // run coroutines after request is finished
             }

--- a/test-util-android/src/main/java/org/fnives/test/showcase/android/testutil/synchronization/idlingresources/IdlingResourcesHelper.kt
+++ b/test-util-android/src/main/java/org/fnives/test/showcase/android/testutil/synchronization/idlingresources/IdlingResourcesHelper.kt
@@ -6,7 +6,7 @@ import org.fnives.test.showcase.android.testutil.synchronization.loopMainThreadF
 import java.util.concurrent.Executors
 
 // workaround, issue with idlingResources is tracked here https://github.com/robolectric/robolectric/issues/4807
-fun anyResourceIdling(): Boolean = !IdlingRegistry.getInstance().resources.all(IdlingResource::isIdleNow)
+fun anyResourceNotIdle(): Boolean = (!IdlingRegistry.getInstance().resources.all(IdlingResource::isIdleNow))
 
 fun awaitIdlingResources() {
     val idlingRegistry = IdlingRegistry.getInstance()
@@ -30,7 +30,7 @@ fun awaitIdlingResources() {
     executor.shutdown()
 }
 
-private fun IdlingResource.awaitUntilIdle() {
+fun IdlingResource.awaitUntilIdle() {
     // using loop because some times, registerIdleTransitionCallback wasn't called
     while (true) {
         if (isIdleNow) return

